File: credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long num = get_long("Number: ");
    long temp = num;
    int sum = 0, count = 0;

    // Luhn’s Algorithm
    while (temp > 0)
    {
        int digit = temp % 10;

        if (count % 2 == 1)
        {
            int dbl = digit * 2;
            sum += (dbl / 10) + (dbl % 10);
        }
        else
        {
            sum += digit;
        }

        temp /= 10;
        count++;
    }

    // Check validity
    if (sum % 10 != 0)
    {
        printf("INVALID\n");
        return 0;
    }

    // Extract first two digits
    long check = num;
    while (check >= 100)
    {
        check /= 10;
    }
    int first2 = check;

    // Decide card type
    if ((first2 == 34 || first2 == 37) && count == 15)
    {
        printf("AMEX\n");
    }
    else if (first2 >= 51 && first2 <= 55 && count == 16)
    {
        printf("MASTERCARD\n");
    }
    else if ((first2 / 10 == 4) && (count == 13 || count == 16))
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

File: WEEK 6/credit.py
# credit card checker using Luhn

num = input("Number: ")

# reverse for easier processing
rev = num[::-1]
total = 0

for i, d in enumerate(rev):
    n = int(d)
    if i % 2 == 1:
        n *= 2
        if n > 9:
            n -= 9
    total += n

if total % 10 != 0:
    print("INVALID")
else:
    # card type check
    if num.startswith("4") and len(num) in [13, 16]:
        print("VISA")
    elif num[:2] in ["34", "37"] and len(num) == 15:
        print("AMEX")
    elif num[:2] in ["51", "52", "53", "54", "55"] and len(num) == 16:
        print("MASTERCARD")
    else:
        print("INVALID")
